client.go

```diff
@@ -1577,17 +1589,65 @@ func dialAddr(addr string, dial DialFunc, dialDualStack, isTLS bool, tlsConfig *
 		}
 		addr = addMissingPort(addr, isTLS)
 	}
-	conn, err := dial(addr)
+	rawConn, err := dial(addr)
 	if err != nil {
 		return nil, err
 	}
-	if conn == nil {
+	if rawConn == nil {
 		panic("BUG: DialFunc returned (nil, nil)")
 	}
 	if isTLS {
-		conn = tls.Client(conn, tlsConfig)
+		var (
+			errChannel chan error
+			timer      *time.Timer
+		)
+
+		if timeout != 0 {
+			errChannel = make(chan error, 2)
```

Comment:
Can you use something like `errorChPool` here to keep the number of allocation in `dialAddr` to a minimum.

Reply:
But I am not sure it's possible in the way it is done with chanel of capacity 1, channel with capacity of 2 should be drained first to be reused

Reply:
Done